### ISSUE TITLE
[Question] Wrong usage? wrong document? or bug? About varBinds of bulkCmd in asynccore/asyncio

### ISSUE THREAD

**Issue by lambdalisue** [etingof/pysnmp#55]
I'm not really sure if my usage is wrong/ the document is wrong/ or a bug. 

The document said

> varBinds (tuple): A sequence of sequences (e.g. 2-D array) of ObjectType class instances representing a table of MIB variables returned in SNMP response. Inner sequences represent table rows and ordered exactly the same as varBinds in request. Number of rows might be less or equal to maxRepetitions value in request.

http://pysnmp.sourceforge.net/docs/hlapi/asyncore/manager/cmdgen/bulkcmd.html#getbulk-command
http://pysnmp.sourceforge.net/docs/hlapi/asyncio/manager/cmdgen/bulkcmd.html#getbulk-command

So I thought the order of appearance of `ObjectType` in a returned `varBinds` is equal to the one in an argument. 
Additionally I found [a code to make `lexicographicMode` in sync-mode](https://github.com/etingof/pysnmp/blob/master/pysnmp/hlapi/asyncore/sync/cmdgen.py#L592) so I tried to make `lexicographicMode=False` like feature with `asyncio` like:

```python
# varBinds := [ObjectType('mib-2.4.22.1.2'), ObjectType('mib-2.4.22.1.3')]
# result = Result(*await bulkCmd(..., *varBinds))
# ...
        for row in result.varBinds:
            print([x[0].prettyPrint() for x in row])
            for c, col in enumerate(row):
                if not objectIdentities[c].isPrefixOf(col[0]):
                    print('Escaped from the scope', col)
```

But actually the order of appearance of `ObjectType` in rows were different from my expectation

```
['SNMPv2-SMI::mib-2.4.22.1.2.27.10.10.0.3', 'SNMPv2-SMI::mib-2.4.22.1.3.27.10.10.0.3']
['SNMPv2-SMI::mib-2.4.22.1.2.27.10.10.0.101', 'SNMPv2-SMI::mib-2.4.22.1.2.27.10.10.0.102']
Escaped from the scope SNMPv2-SMI::mib-2.4.22.1.2.27.10.10.0.102 = 0x784f43835a20
['SNMPv2-SMI::mib-2.4.22.1.2.27.10.10.0.103', 'SNMPv2-SMI::mib-2.4.22.1.2.27.10.10.0.106']
Escaped from the scope SNMPv2-SMI::mib-2.4.22.1.2.27.10.10.0.106 = 0x6480994d9e54
['SNMPv2-SMI::mib-2.4.22.1.2.27.10.10.0.107', 'SNMPv2-SMI::mib-2.4.22.1.2.27.10.10.0.108']
Escaped from the scope SNMPv2-SMI::mib-2.4.22.1.2.27.10.10.0.108 = 0x00220c23c840
...
Done asynccore
['SNMPv2-SMI::mib-2.4.22.1.2.27.10.10.0.3', 'SNMPv2-SMI::mib-2.4.22.1.3.27.10.10.0.3']
['SNMPv2-SMI::mib-2.4.22.1.2.27.10.10.0.101', 'SNMPv2-SMI::mib-2.4.22.1.2.27.10.10.0.102']
Escaped from the scope SNMPv2-SMI::mib-2.4.22.1.2.27.10.10.0.102 = 0x784f43835a20
['SNMPv2-SMI::mib-2.4.22.1.2.27.10.10.0.103', 'SNMPv2-SMI::mib-2.4.22.1.2.27.10.10.0.106']
Escaped from the scope SNMPv2-SMI::mib-2.4.22.1.2.27.10.10.0.106 = 0x6480994d9e54
...
Done asyncio
```

So I wonder if my usage is wrong or the document is wrong, or bug.

Thanks for reading 😄 

---

The entire code I used is

```python
from collections import namedtuple
from pysnmp.hlapi.varbinds import CommandGeneratorVarBinds

IP_NET_TO_MEDIA_PHYS_ADDRESS = '1.3.6.1.2.1.4.22.1.2'
IP_NET_TO_MEDIA_NET_ADDRESS = '1.3.6.1.2.1.4.22.1.3'

Result = namedtuple('Result', [
    'errorIndication',
    'errorStatus',
    'errorIndex',
    'varBinds',
])


def with_asynccore():
    from pysnmp.hlapi import asyncore as pysnmp

    snmpEngine = pysnmp.SnmpEngine()
    communityData = pysnmp.CommunityData('public')
    transportTarget = pysnmp.UdpTransportTarget(('10.10.0.1', 161))
    objectTypes = [
        pysnmp.ObjectType(pysnmp.ObjectIdentity(IP_NET_TO_MEDIA_PHYS_ADDRESS)),
        pysnmp.ObjectType(pysnmp.ObjectIdentity(IP_NET_TO_MEDIA_NET_ADDRESS)),
    ]
    vbproc = CommandGeneratorVarBinds()
    objectIdentities = [
        x[0] for x in vbproc.makeVarBinds(snmpEngine, objectTypes)
    ]

    def cbFun(snmpEngine, sendRequestHandle, errorIndication, errorStatus,
              errorIndex, varBinds, cbCtx):
        for row in varBinds:
            print([x[0].prettyPrint() for x in row])
            for c, col in enumerate(row):
                if not objectIdentities[c].isPrefixOf(col[0]):
                    print('Escaped from the scope', col)

    snmpEngine = pysnmp.SnmpEngine()
    pysnmp.bulkCmd(snmpEngine,
                   communityData,
                   transportTarget,
                   pysnmp.ContextData(),
                   0, 10,
                   *objectTypes,
                   cbFun=cbFun)
    snmpEngine.transportDispatcher.runDispatcher()
    print('Done asynccore')


def with_asyncio():
    import asyncio
    from pysnmp.hlapi import asyncio as pysnmp

    snmpEngine = pysnmp.SnmpEngine()
    communityData = pysnmp.CommunityData('public')
    transportTarget = pysnmp.UdpTransportTarget(('10.10.0.1', 161))
    objectTypes = [
        pysnmp.ObjectType(pysnmp.ObjectIdentity(IP_NET_TO_MEDIA_PHYS_ADDRESS)),
        pysnmp.ObjectType(pysnmp.ObjectIdentity(IP_NET_TO_MEDIA_NET_ADDRESS)),
    ]
    vbproc = CommandGeneratorVarBinds()
    objectIdentities = [
        x[0] for x in vbproc.makeVarBinds(snmpEngine, objectTypes)
    ]

    async def run(varBinds):
        result = Result(*await pysnmp.bulkCmd(snmpEngine,
                                              communityData,
                                              transportTarget,
                                              pysnmp.ContextData(),
                                              0, 10,
                                              *varBinds))
        for row in result.varBinds:
            print([x[0].prettyPrint() for x in row])
            for c, col in enumerate(row):
                if not objectIdentities[c].isPrefixOf(col[0]):
                    print('Escaped from the scope', col)

        snmpEngine.transportDispatcher.runDispatcher()

    asyncio.get_event_loop().run_until_complete(run(objectTypes))
    print('Done asyncio')

with_asynccore()

with_asyncio()
```

**Comment by etingof** [comment]
Tl;DR; I tend to blame your SNMP agent for returning you wrong OIDs.

Regardless of I/O transport being used, pysnmp returns SNMP tables as a 2-D array of OID-value pairs. Let's say you run GETNEXT or GETBULK query for imaginary OIDs:

    1.2.3   1.3.4

For as long as the agent can serve these (and assuming non-repetitions parameter is set to 0), the GETBULK would return a 2-D array like:

    1.2.4  1.3.5
    1.2.5  1.3.6
    1.2.6  1.3.7
     .....

In other words, for each request OID the agent should serve the ever increasing column of greater OIDs.

There may be a confusion though -- by some point the next OID at one column may overlap with the sequence of OIDs at the other column:

    1.2.3  1.2.4
    1.2.4  1.2.5
     ....

What I can see in your output it surprising -- it looks like your agent is returning wrong OIDs. If you point your script to `demo.snmplabs.com`, the output would look consistently with my description:

    ['SNMPv2-SMI::mib-2.4.22.1.2.2.195.218.254.97', 'SNMPv2-SMI::mib-2.4.22.1.3.2.195.218.254.97']
    ['SNMPv2-SMI::mib-2.4.22.1.3.2.195.218.254.97', 'SNMPv2-SMI::mib-2.4.22.1.4.2.195.218.254.97']
    ['SNMPv2-SMI::mib-2.4.22.1.4.2.195.218.254.97', 'SNMPv2-SMI::mib-2.4.23.0']
    ['SNMPv2-SMI::mib-2.4.23.0', 'SNMPv2-SMI::mib-2.4.24.4.1.1.0.0.0.0.0.0.0.0.0.195.218.254.97']
    ['SNMPv2-SMI::mib-2.4.24.4.1.1.0.0.0.0.0.0.0.0.0.195.218.254.97', 'SNMPv2-SMI::mib-2.4.24.4.1.1.127.0.0.0.0.0.0.255.0.0.0.0.0']
    ['SNMPv2-SMI::mib-2.4.24.4.1.1.127.0.0.0.0.0.0.255.0.0.0.0.0', 'SNMPv2-SMI::mib-2.4.24.4.1.1.195.218.254.0.0.255.255.255.0.0.0.0.0']
    ['SNMPv2-SMI::mib-2.4.24.4.1.1.195.218.254.0.0.255.255.255.0.0.0.0.0', 'SNMPv2-SMI::mib-2.4.24.4.1.2.0.0.0.0.0.0.0.0.0.195.218.254.97']
    ['SNMPv2-SMI::mib-2.4.24.4.1.2.0.0.0.0.0.0.0.0.0.195.218.254.97', 'SNMPv2-SMI::mib-2.4.24.4.1.2.127.0.0.0.0.0.0.255.0.0.0.0.0']
    ['SNMPv2-SMI::mib-2.4.24.4.1.2.127.0.0.0.0.0.0.255.0.0.0.0.0', 'SNMPv2-SMI::mib-2.4.24.4.1.2.195.218.254.0.0.255.255.255.0.0.0.0.0']
    ['SNMPv2-SMI::mib-2.4.24.4.1.2.195.218.254.0.0.255.255.255.0.0.0.0.0', 'SNMPv2-SMI::mib-2.4.24.4.1.3.0.0.0.0.0.0.0.0.0.195.218.254.97']

What do you think? We can debug this issue further if you wish.

**Comment by etingof** [comment]
I am closing this issue as it idles for quite long. Feel free to re-open it if you think it needs further attention.
